gateway-api: Support GRPCRoute resource

sayboras opened the pull request:
### Description

Please refer to individual commits for more details.

Fixes: https://github.com/cilium/cilium/issues/21928

```release-note
gateway-api: Support GRPCRoute resource
```

sayboras committed:
gateway-api: Add no-op grpcRoute reconciler

This commit is to add no-op grpcRoute reconciler, which will kick off
for one for the below changes:

- Any change for GRPCRoute objects
- Any change for backend services used by any GRPCRoute object
- Any change for related Reference Grant
- Any change for parent Gateway

The logic is pretty much similar to what we have with HTTPRoute
reconciler.

Signed-off-by: Tam Mach <tam.mach@cilium.io>

sayboras committed:
gateway-api: Add route checks for GRPCRoute

This is to follow the same Input validation interface, which is used to
perform all the checks for HTTPRoute and TLSRoute.

Signed-off-by: Tam Mach <tam.mach@cilium.io>

sayboras committed:
gateway-api: Refactor ingestion method

Just lift and shift some code blocks for readability, there is no change
in logic at all.

Signed-off-by: Tam Mach <tam.mach@cilium.io>

sayboras committed:
gateway-api: Add ingestion logic for GRPCRoute

Most of the logic is same as HTTPRoute, just different data types and
structs make it hard to generalize.

Signed-off-by: Tam Mach <tam.mach@cilium.io>

sayboras committed:
gateway-api: Support GRPCRoute in Gateway

This commit is to add steps to configure related accepted GRPCRoute for
parent Gateway.

Signed-off-by: Tam Mach <tam.mach@cilium.io>

sayboras committed:
helm: Grant grpcroute permissions to operator

Signed-off-by: Tam Mach <tam.mach@cilium.io>

sayboras committed:
docs: Add GRPCRoute support in gateway api

Signed-off-by: Tam Mach <tam.mach@cilium.io>

sayboras committed:
gateway-api: Add grpc_web and grpc_stats filters

Signed-off-by: Tam Mach <tam.mach@cilium.io>

sayboras committed:
gha: Add GRPCRoute CRD in CI

Signed-off-by: Tam Mach <tam.mach@cilium.io>

sayboras committed:
gateway-api: Avoid protocol error for grpc web

Use explicit http config in envoy cluster to avoid the below issue when
trying to send curl command to GRPC endpoint.

Before

```
< HTTP/1.1 502 Bad Gateway
< content-length: 0
< content-type: application/grpc-web-text+proto
< grpc-message: upstream connect error or disconnect/reset before headers. reset reason: protocol error
< grpc-status: 14
< date: Tue, 17 Oct 2023 13:57:43 GMT
< server: envoy
```

After
```
< HTTP/1.1 200 OK
< content-type: application/grpc-web-text+proto
< x-envoy-upstream-service-time: 0
< date: Tue, 17 Oct 2023 13:57:54 GMT
< server: envoy
< transfer-encoding: chunked
```

Signed-off-by: Tam Mach <tam.mach@cilium.io>

qmonnet reviewed (approved):
I came for the docs, and the docs looked good.

nbusseneau reviewed (approved):
Trivial ack for ci-structure.

squeed commented on operator/pkg/gateway-api/grpcroute.go:
```diff
@@ -0,0 +1,169 @@
+		return err
+	}
+
+	if err := mgr.GetFieldIndexer().IndexField(context.Background(), &gatewayv1alpha2.GRPCRoute{}, gatewayIndex,
```
This line is is some proper arcane incantation. Can you please add a comment :-)?

squeed commented on operator/pkg/gateway-api/grpcroute_reconcile.go:
```diff
@@ -0,0 +1,132 @@
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
```
Please add a comment giving a basic overview, especially the inputs and outputs.

mhofstetter commented on .github/workflows/conformance-gateway-api.yaml:
```diff
@@ -169,6 +169,7 @@ jobs:
           kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/gateway-api/${{ env.gateway_api_version }}/config/crd/experimental/gateway.networking.k8s.io_tlsroutes.yaml
+          kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/gateway-api/${{ env.gateway_api_version }}/config/crd/experimental/gateway.networking.k8s.io_grpcroutes.yaml
 
           # To make sure that Gateway API CRs are available
```
shouldn't we add a check for the `GrpcRoute`?

sayboras replied:
ah good catch :100:

mhofstetter commented on operator/pkg/gateway-api/grpcroute.go:
```diff
@@ -0,0 +1,169 @@
+
+	if err := mgr.GetFieldIndexer().IndexField(context.Background(), &gatewayv1alpha2.GRPCRoute{}, gatewayIndex,
+		func(rawObj client.Object) []string {
+			hr := rawObj.(*gatewayv1alpha2.GRPCRoute)
```
nit: `hr` -> `gr` (or just `route`)

mhofstetter commented on operator/pkg/gateway-api/grpcroute.go:
```diff
@@ -0,0 +1,169 @@
+// SetupWithManager sets up the controller with the Manager.
+func (r *grpcRouteReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	if err := mgr.GetFieldIndexer().IndexField(context.Background(), &gatewayv1alpha2.GRPCRoute{}, backendServiceIndex,
+		func(rawObj client.Object) []string {
```
might be worth to define the index functions as normal functions (and reference them here) instead of anonymous functions. should be easier to read and document them.

joamaki reviewed (approved):
Approve for `install/kubernetes` changes.

mhofstetter commented on operator/pkg/gateway-api/grpcroute.go:
```diff
@@ -0,0 +1,169 @@
+func (r *grpcRouteReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	if err := mgr.GetFieldIndexer().IndexField(context.Background(), &gatewayv1alpha2.GRPCRoute{}, backendServiceIndex,
+		func(rawObj client.Object) []string {
+			hr, ok := rawObj.(*gatewayv1alpha2.GRPCRoute)
```
nit: `hr` -> `gr` (or just `route`)

mhofstetter commented on operator/pkg/gateway-api/gateway_reconcile.go:
```diff
@@ -226,6 +233,16 @@ func (r *gatewayReconciler) filterHTTPRoutesByGateway(ctx context.Context, gw *g
 	return filtered
 }
 
+func (r *gatewayReconciler) filterGRPCRoutesByGateway(ctx context.Context, gw *gatewayv1beta1.Gateway, routes []gatewayv1alpha2.GRPCRoute) []gatewayv1alpha2.GRPCRoute {
```
maybe out of scope for this PR: we should start thinking about a way to get rid of this boilerplate for every route-type.

sayboras replied:
yeah, we might need to wrap the types, similar to what we have with input check. 

Will do this as a follow up.

mhofstetter commented on .github/workflows/conformance-gateway-api.yaml:
```diff
@@ -169,6 +169,7 @@ jobs:
           kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/gateway-api/${{ env.gateway_api_version }}/config/crd/${{ matrix.crd-channel }}/gateway.networking.k8s.io_referencegrants.yaml
           ## TLSRoute is only available in experimental channel in v0.7.0
           kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/gateway-api/${{ env.gateway_api_version }}/config/crd/experimental/gateway.networking.k8s.io_tlsroutes.yaml
+          kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/gateway-api/${{ env.gateway_api_version }}/config/crd/experimental/gateway.networking.k8s.io_grpcroutes.yaml
```
do we have to add the grpcRoute to the list of supported features to include this in the gateway-api conformance tests?

https://github.com/cilium/cilium/blob/7add21c7e44386cb5d9f72954bf3b911e4ce242a/.github/workflows/conformance-gateway-api.yaml#L121

edit: or is this the gRPC Route not yet covered by the conformance tests or included in the http route? :thinking: https://gateway-api.sigs.k8s.io/geps/gep-917/#conformance-profiles

sayboras replied:
yup, there is no conformance test right now, however, the translation logic is well covered in HTTP conformance tests, which gives us high enough confidence :pray:

mhofstetter replied:
But i assume the HTTP conformance tests don't make use of the CRD `GRPCRoute`? if not we should at least introduce such a test (e.g. in the connectivity tests) as a follow up. WDYT?

sayboras replied:
yup, makes sense to add some basic grpcroute test instead of doing it manually :+1:

sayboras replied:
let me do it in this PR as well, marking this draft to add grpc test

sayboras replied:
done, one simple grpc echo route, and curl with grpc web steps are added.

mhofstetter replied:
nice, thanks!

mhofstetter reviewed (commented):
nice work @sayboras :partying_face: 

i took a first look and added some comments inline. I'm not sure whether the gateway-api conformance tests are already enabled without having added the grpcRoute feature to the list of supported features. Might be worth to activate that first and have some feedback from the conformance tests too :smile: (comment: https://github.com/cilium/cilium/pull/28654#discussion_r1365137956)

sayboras committed:
gha: Remove redundant Cilium GatewayClass creation

This should be created as part of recent version of Cilium CLI

Signed-off-by: Tam Mach <tam.mach@cilium.io>

sayboras committed:
gha: Add simple GRPCRoute test

This is to add basic GPRCRoute test, till we have more extensive
conformance test from upstream.

Signed-off-by: Tam Mach <tam.mach@cilium.io>

mhofstetter reviewed (approved):
LGTM, thanks again!
